File: April/Abiji-2020/Golang /Reveal Cards In Increasing Order/Reveal Cards In Increasing order.go

type deque []int

func (d *deque) pushBack(val int) {
	*d = append(*d, val)
}

func (d *deque) popFront() int {
	front := (*d)[0]
	*d = (*d)[1:]
	return front
}

func deckRevealedIncreasing(deck []int) []int {
	ans := make([]int, len(deck))
	dq := make(deque, 0)
	for i := 0; i < len(deck); i++ {
		dq.pushBack(i)
	}
	sort.Ints(deck)
	for i := 0; i < len(deck); i++ {
		card := deck[i]
		front := dq.popFront()
		ans[front] = card
		if len(dq) > 0 {
			dq.pushBack(dq.popFront())
		}
	}
	return ans
}

File: April/Abiji-2020/C++/Reveal Cards in Increasing order/Reveal Cards in increasing order.cpp
class Solution {
public:
    vector<int> deckRevealedIncreasing(vector<int>& deck) {
        vector<int> ans(deck.size(),0);
        deque<int> dq;
        for(int i=0;i<deck.size();i++){
            dq.push_back(i);
        }
        sort(deck.begin(),deck.end());
        for(int i=0;i<deck.size();i++){
            int card = deck[i];
            ans[dq.front()] = card;
            dq.pop_front();
            if(!dq.empty()){
                dq.push_back(dq.front());
                dq.pop_front();
            } 
        }
        return ans;
    }
};
